### PR TITLE
Add custom slash command argument converters

### PR DESCRIPTION
**Pull request by IDoEverything**
# Summary
Adds the ability for slash command users to define their own argument converters. Also moves all the default converters to the same system.

# Details
Uses a similar system to how it works in commandsnext. This greatly cleans up the slashcommandsextension.cs file, and makes a distinction between the converters for the default values and the converters for timespan and discordemoji which aren't parsed by discord.

### CODE REVIEW

File: DSharpPlus.SlashCommands/Converters/CustomConverters.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+namespace DSharpPlus.SlashCommands.Converters
+{
+    public class TimespanConverter : ISlashArgumentConverter<TimeSpan?>
+    {
+        public ApplicationCommandOptionType OptionType { get; } = ApplicationCommandOptionType.String;
+
+        public Task<TimeSpan?> Convert(DiscordInteractionDataOption value, InteractionContext context)
+        {
+            var timeSpanRegex =
+                new Regex(@"^(?<days>\d+d\s*)?(?<hours>\d{1,2}h\s*)?(?<minutes>\d{1,2}m\s*)?(?<seconds>\d{1,2}s\s*)?$",
```
Comment (VelvetThePanda) [inline DSharpPlus.SlashCommands/Converters/CustomConverters.cs:39]: Regex is relatively expensive to initialize; you should be putting this into a (static) readonly field

File: DSharpPlus.SlashCommands/Converters/CustomConverters.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+            if (int.TryParse(value.Value.ToString(), NumberStyles.Number, CultureInfo.InvariantCulture, out _))
+            {
+                return Task.FromResult((TimeSpan?)null);
+            }
+
+            if (TimeSpan.TryParse(value.Value.ToString().ToLowerInvariant(), CultureInfo.InvariantCulture, out var result))
+            {
+                return Task.FromResult((TimeSpan?)result);
+            }
```
Comment (VelvetThePanda) [inline DSharpPlus.SlashCommands/Converters/CustomConverters.cs:54]: Consistency: Braces can/should be removed here

File: DSharpPlus.SlashCommands/Converters/CustomConverters.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                    case 's':
+                        s = val;
+                        break;
+                }
+            }
+
+            result = new TimeSpan(d, h, m, s);
+            return Task.FromResult((TimeSpan?)result);
+        }
+    }
```
Comment (VelvetThePanda) [inline DSharpPlus.SlashCommands/Converters/CustomConverters.cs:100]: I have a micro ts parser somewhere around here if you'd like to use that,

File: DSharpPlus.SlashCommands/Converters/DefaultConverters.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    public class BoolConverter : ISlashArgumentConverter<bool>
+    {
+        public ApplicationCommandOptionType OptionType { get; } = ApplicationCommandOptionType.Boolean;
+        public Task<bool> Convert(DiscordInteractionDataOption value, InteractionContext context)
+            => Task.FromResult((bool)value.Value);
+    }
+
+    public class NullableBoolConverter : ISlashArgumentConverter<bool?>
```
Comment (VelvetThePanda) [inline DSharpPlus.SlashCommands/Converters/DefaultConverters.cs:61]: I think what should happen here (though this probably isbn't possible given the way you've written this) is that you should parse for nullable generic structs (`<T?> where T : struct`), and attempt to grab the parser for that type instead.

File: DSharpPlus.SlashCommands/Converters/DefaultConverters.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    public class AttachmentConverter : ISlashArgumentConverter<DiscordAttachment>
+    {
+        public ApplicationCommandOptionType OptionType { get; } = ApplicationCommandOptionType.Attachment;
+        public Task<DiscordAttachment> Convert(DiscordInteractionDataOption value, InteractionContext context)
+        {
+            if (context.Interaction.Data.Resolved.Attachments?.ContainsKey((ulong)value.Value) ?? false)
+            {
+                var attachment = context.Interaction.Data.Resolved.Attachments[(ulong)value.Value];
+                return Task.FromResult(attachment);
+            }
```
Comment (VelvetThePanda) [inline DSharpPlus.SlashCommands/Converters/DefaultConverters.cs:154]: Nitpick: Consider initializing a variable for clarity instead

File: DSharpPlus.SlashCommands/SlashCommandsExtension.cs [hunk 6]
```diff
@@ -479,33 +509,28 @@ private static List<DiscordApplicationCommandOptionChoice> GetChoiceAttributesFr
...
-                return ApplicationCommandOptionType.Mentionable;
+            var type = typeof(T);
             if (type.IsEnum || Nullable.GetUnderlyingType(type)?.IsEnum == true)
                 return ApplicationCommandOptionType.String;
-            if (type == typeof(DiscordAttachment))
-                return ApplicationCommandOptionType.Attachment;
-            throw new ArgumentException("Cannot convert type! Argument types must be string, long, bool, double, TimeSpan?, DiscordChannel, DiscordUser, DiscordRole, DiscordEmoji, DiscordAttachment, SnowflakeObject, or an Enum.");
+            if (!ArgumentConverters.Any(converter => converter.Key == type))
+                throw new ArgumentException(
+                    @$"Could not convert type of argument. Provided {type.Name}, valid argument types are: {string.Join(", ", ArgumentConverters.Select(c => c.Key.IsGenericType &&
```
Comment (VelvetThePanda) [inline DSharpPlus.SlashCommands/SlashCommandsExtension.cs:519]: This line is particularly long; consider using braces and initializing some variables

File: DSharpPlus.SlashCommands/SlashCommandsExtension.cs [hunk 7]
```diff
@@ -837,154 +862,37 @@ private async Task<List<object>> ResolveInteractionCommandParameters(Interaction
...
-                            this.Client.Logger.LogError("Missing attachment in resolved data. This is an issue with Discord.");
+                        args.Add(Enum.Parse(Nullable.GetUnderlyingType(parameter.ParameterType), (string)option.Value));
+                        continue;
                     }
 
-                    else
-                        throw new ArgumentException("Error resolving interaction.");
+                    //Convert parameter using registered converter.
+                    var converted = await (typeof(SlashCommandsExtension).GetTypeInfo().DeclaredMethods
+                        .First(m => m.Name == nameof(this.ConvertArgument)).MakeGenericMethod(parameter.ParameterType).Invoke(this, new object[] {option, context}) as Task<object>).ConfigureAwait(false);
```
Comment (VelvetThePanda) [inline DSharpPlus.SlashCommands/SlashCommandsExtension.cs:880]: Consider adding some newlines for readability.

**Review by VelvetThePanda — CHANGES_REQUESTED** [review]

File: DSharpPlus.SlashCommands/Converters/DefaultConverters.cs (hunk 4 above)
Comment (IDoEverything) [reply DSharpPlus.SlashCommands/Converters/DefaultConverters.cs]: This has been fixed.

**Comment by OoLunar** [comment]
Is this PR functioning correctly and ready to be merged?

**Comment by VelvetThePanda** [comment]
Superseded by #1362

**Pull request closed without merging**